### PR TITLE
Add homogeneous tuples in C

### LINKED ISSUE #850: Homogeneous tuples in C
**Describe the bug**
Homogeneous tuples cannot be implemented in the same way as inhomogeneous tuples as they have an additional difficulty. They must be indexable and the size may be unknown. However they are not as complicated as arrays. They cannot be poiner targets and cannot be modified so there is no need to handle slices or pointers

**Test Code**
Provide code which does not currently work but which should do when this issue is fixed:
```python
a = (1,2,3)
```

**Expected Behaviour**
Mimicking, the solution in Fortran, we should use the ndarray structure and treat a tuple containing objects of different shapes as inhomogeneous

**Language**
C

### PR DESCRIPTION
**Pull request by EmilyBourne**
Add homogeneous tuples in C. Fixes #850 . Fix bugs in unravelling. Improve python printer

**Commit Summary**
- When substituting add new user node before removing old user node to avoid unnecessary node invalidation
- Add `clear_user_nodes` function to prevent objects holding their syntactic and semantic users when they are valid for both stages (e.g. Literals)
- Add simplification step to `PyccelAdd` avoid printing `a+0`
- Unravel expressions such as `a[1:3] = (4,5)` or `a[:] = np.array([3,4])` these expressions don't require memory allocation and have known size
- Correct `Duplicate` loop creation (fixed wrong size, collect multiple expressions to handle tuple assignment unravelling)
- Update `expand_inhomog_tuple_assignments` doc
- Ensure `Duplicate`, `Concatenate`, and `HomogeneousTupleVariable` assignments have memory allocation before unravelling (accessing elements)
- Add `__iter__` to `HomogeneousTupleVariable` (default `__iter__` calls `__getitem__` until `IndexError` is raised, but we don't always know the size)
- Allow `VariableAddress` to contain `IndexedElement`
- Allow `copy_NumpyArray_Data` to also handle homogenous `PythonTuple`s
- Make tuples of arrays inhomogeneous (neither our C framework, nor fortran can handle arrays of pointers)
- Activate tuple tests
- Add functions in functions to python printer
- Add `_print_Duplicate` and `_print_Concatenate` to python printer
- Correct tuple indexing in python printer to print `a[i][j]` instead of `a[i,j]` when indexing `((1,2),(3,4))`
- Avoid printing `from pyccel.decorators import types` multiple times
- Add `_print_Interface` (to be improved, see #885) to python printer

### CODE REVIEW

File: pyccel/parser/semantic.py [hunk 1]
```diff
@@ -1446,18 +1449,25 @@ def _visit_CodeBlock(self, expr, **settings):
...
         return expr
     def _visit_Comment(self, expr, **settings):
+        expr.clear_user_nodes()
         return expr
     def _visit_CommentBlock(self, expr, **settings):
+        expr.clear_user_nodes()
         return expr
     def _visit_AnnotatedComment(self, expr, **settings):
+        expr.clear_user_nodes()
         return expr
```
Comment (yguclu) [inline pyccel/parser/semantic.py:1471]: Could you please separate these methods from each other with an empty line?

File: tests/epyccel/modules/tuples.py [hunk 2]
```diff
@@ -88,7 +88,9 @@ def inhomogenous_tuple_3():
     return ai[0], ai[1], ai[2]
 
 def inhomogenous_tuple_2_levels_1():
-    ai = ((1,2), (4,False), (3.0, 'boo'))
+    # TODO [EB 15.06.21] Put back original test when strings are supported
+    #ai = ((1,2), (4,False), (3.0, 'boo'))
+    ai = ((1,2), (4,False), (3.0, True))
```
Comment (yguclu) [inline tests/epyccel/modules/tuples.py:93]: How was the original test working?

**Review by yguclu — CHANGES_REQUESTED** [review]
Excellent work! I only have two comments

File: tests/epyccel/modules/tuples.py (hunk 2 above)
Comment (EmilyBourne) [reply tests/epyccel/modules/tuples.py:93]: Fortran supports strings. As this is an inhomogeneous tuple it causes no problems in fortran, but strings aren't supported in C at all

File: pyccel/parser/semantic.py (hunk 1 above)
Comment (EmilyBourne) [reply pyccel/parser/semantic.py]: Fixed. See c62542a

Comment (yguclu) [reply pyccel/parser/semantic.py]: Thanks

File: tests/epyccel/modules/tuples.py (hunk 2 above)
Comment (yguclu) [reply tests/epyccel/modules/tuples.py:93]: Maybe we should write "when strings are supported **in C**" then...

Comment (EmilyBourne) [reply tests/epyccel/modules/tuples.py]: Fixed 62ec0d7

**Review by yguclu — APPROVED** [review]

**Pull request merged**